File: clojure/transpose/src/transpose.clj
(ns transpose (:require [clojure.string :as str]))

(defn- max-width [lines] (if (seq lines) (apply max (map count lines)) 0))

(defn- last-row-with-column
  [lines col]
  (last (keep-indexed (fn [row line] (when (< col (count line)) row)) lines)))

(defn- char-at [line col] (if (< col (count line)) (nth line col) \space))

(defn transpose
  "Returns the transposed version of the given string."
  [s]
  (let [lines (str/split-lines s)
        width (max-width lines)
        transposed-lines (for [col (range width)]
                           (let [last-row (last-row-with-column lines col)]
                             (apply str (for [row (range (inc last-row))]
                                          (char-at (nth lines row) col)))))]
    (str/join "\n" transposed-lines)))

File: python/transpose/transpose.py
"""Transpose text"""

# pylint: disable=missing-function-docstring

from itertools import zip_longest

def transpose(text):
    def transpose_row(row):
        last = max(
            (index for index, char in enumerate(row) if char is not None),
            default=-1
        )
        return ''.join(' ' if char is None else char for char in row[:last + 1])
    
    return '\n'.join(
        transpose_row(row)
        for row in zip_longest(*text.split('\n'), fillvalue=None)
    )
